### ISSUE TITLE
Meshcentral Nginx Unable to connect websocket

### ISSUE THREAD

**Issue by Alsterdetektive1** [Ylianst/MeshCentral#3762]
I installed MeshCentral via npm on my Server (Ubuntu 20.04). Now i am trying to reverse proxy it via nginx. I followed the instruction and was able to make the admin account, but after trying to login the "Unable to connect web socket, click to reconnect" popped up.

----
My configuration:

**config.json:**

```
{
  "$schema": "http://info.meshcentral.com/downloads/meshcentral-config-schema.json>
  "__comment1__": "This is a simple configuration file, all values and sections>
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for>
  "settings": {
    "cert": "meshcentral.mywebsite.de",
    "WANonly": true,
    "LANonly": false,
    "_sessionKey": "MyReallySecretPassword1",
    "Port": 9006,
    "aliasPort": 9006,
    "redirPort": 9007,
    "redirAliasPort": 9007,
    "TlsOffload": "127.0.0.1",
    "AgentPong": 300
  },
  "domains": {
    "": {
      "_title": "MyServer",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true,
      "_certUrl": "https://localhost:9006"
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first b>
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "production": false
  }
}
```

----
**Nginx Conf:**

```
server {
        listen 443 ssl;
        server_name _;
        server_name meshcentral.mysite.de;
        ssl_certificate /etc/letsencrypt/live/meshcentral.mysite.de/fullchain.pem>
        ssl_certificate_key /etc/letsencrypt/live/meshcentral.mysite.de/privkey.pem>
        proxy_send_timeout 330s;
        proxy_read_timeout 330s;
        location / {
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection "upgrade";
                proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                proxy_set_header X-Forwarded-Proto $scheme;
                proxy_set_header X-Forwarded-Host $host:$server_port;
                proxy_pass http://localhost:9006/;
        }
        error_log /var/log/nginx/meshcentral.error.log;
        access_log /var/log/nginx/meshcentral.access.log;
```

----
I have no clue what I am doing wrong, but it doesnt work.
Thanks in advance

**Comment by vitko-bg** [comment]
Try to add in domain section on your config.json the following line

"CertUrl": "https://127.0.0.1:443"

That is what I have in my config and all works fine.

**Comment by Alsterdetektive1** [comment]
Sadly changes nothing, error is still there

**Comment by vitko-bg** [comment]
> 

Did you restart MC server after changes are made?

**Comment by Alsterdetektive1** [comment]
Yes, made the systemd script and restarted that, and restarted nginx too

**Comment by Alsterdetektive1** [comment]
and log is also printing "Loaded web certificate from 
```
https://127.0.0.1:443", host: "meshcentral.mysite.de"
    SHA384 cert hash: ...
    SHA384 key hast: ...
MeshCentral HTTP server running on port 9006
```

**Comment by Ylianst** [comment]
For the web socket not connecting, I don't see what could be wrong... your config looks pretty good. Can you open the browser console window and try again, then post any errors you see? NGINX logs may also help figure out what is going on.

I have a [YouTube video on installing MeshCentral with NGINX here](https://www.youtube.com/watch?v=YSmiLyKSX2I), that can hopefully help. You can also look at a sample NGINX config in section 16 of [the User's Guide here](https://meshcentral.com/info/docs/MeshCentral2UserGuide.pdf).

It looks like you want MeshCentral on port 9006, but it will be visible on port 443 externally, so you need:

```
    "Port": 9006,
    "AliasPort": 443,
    "RedirPort": 9007,
    "RedirAliasPort": 80,
```

This indicates, listen on 9006 but assume your on 443 externally for real. This is important when configuring MeshAgents and other things that need to know the external port. Also, in your NGINX config, do redirect HTTP/80 to port 9007. If you don't want to do that, then:

```
    "Port": 9006,
    "AliasPort": 443,
    "RedirPort": 0,
```

In this case, MeshCentral will turn off the redirection server.

The "certurl" is needed, that looks like it's working.

**Comment by OutbackMatt** [comment]
This is my working setup
(Ubuntu 20.04, NginX, LetsEncrypt via certbot)
Perhaps the order of the configuration variables matters

Or perhaps the proxy pass needs the FQDN (and the relevant port forwarded through the router, and opened on the server firewall)

server {
		listen         80;
        server_name mesh.example.com;
 
		return         301 https://$server_name$request_uri;
}

server {
		# SSL configuration
		listen 443 ssl;
#		listen [::]:443 ssl;
        server_name mesh.example.com;

		add_header Strict-Transport-Security "max-age=15552000; includeSubDomains" always;
		ssl_certificate /etc/letsencrypt/live/mesh.example.com/fullchain.pem;
		ssl_certificate_key /etc/letsencrypt/live/mesh.example.com/privkey.pem;
        

		# MeshCentral uses long standing web socket connections, set longer timeouts.
		proxy_send_timeout 330s;
		proxy_read_timeout 330s;

		location / {
			proxy_pass https://mesh.example.com:444/;
			proxy_http_version 1.1;
			proxy_hide_header content-security-policy;
			# Allows websockets over HTTPS.
			proxy_set_header Upgrade $http_upgrade;
			proxy_set_header Connection "upgrade";
            
			proxy_set_header Host $host;

			# Inform MeshCentral about the real host, port and protocol
			proxy_set_header X-Forwarded-Host $host:$server_port;
			proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
			proxy_set_header X-Forwarded-Proto $scheme;
		}
}

**Comment by Alsterdetektive1** [comment]
Thanks for your help.
Changed the AliasPort like told above and changed the nginx settings to nearly the same (just let proxy_pass with http) but sadly nothing changed.
Read yesterday something about old nginx versions maybe broken with node. I use my whole server structure with the nginx of Gitlab-Omnibus do you think this could be the problem?

**Comment by Ylianst** [comment]
I have no idea. It would be interesting to see the NGINX logs to know what is going on. Other than that, I can't tell with the information provided. Your configs look ok, it should work.

**Comment by Alsterdetektive1** [comment]
Okay okay, I found the problem, just because I had another problem with Medusa and BBB.
In this case i wrote 
`proxy_set_header X-Forwarded-Host $host:$server_port;`

The trick is to do it as 2 two line

```
proxy_set_header X-Forwarded-Port $server_port;
proxy_set_header X-Forwarded-Host $host;
```

Thank you all for your help and hopefully we can help another lost soul with the same problem.

**Comment by Alsterdetektive1** [comment]
Okay, nevermind guys, this other nginx configs let it just work for some seconds, after that and clicking a bit around, the error occurs again. I will try to send tomorrow the logs @Ylianst .

**Comment by Ylianst** [comment]
I have no idea what Medusa and BBB are. If you have a idle timeout for websockets in your infrastructure, try this in the settings section of the config.json:

```
"browserPing": 20
```

This will cause dummy traffic to be sent to the browser every 20 seconds.
